docs/developers/talawa/flutter_testing_guide.md

```diff
@@ -0,0 +1,271 @@
+  verify(calculator.add(2, 3)).called(1);
+
+  // Verify that the result returned by the controller is correct
+  expect(result, equals(5));
+});
+  ```
+  
+  In this test, the `MockCalculator` is created and passed to the `CalculatorController`. The `when` method is used to stub the `add` method on the mock calculator, so
+  that when the `add` method is called with arguments 2 and 3, it returns the value 5. Then, the `addNumbers` method is called on the `CalculatorController`, and the
+  result is verified using the `expect` method. Finally, the `verify` method is used to ensure that the `add` method was called on the mock calculator with the correct
+  arguments.
+
+
+## How to Stub Methods and Classes?
+
+Stubbing is a technique used in testing to replace a real object with a simplified version that provides predictable behavior.
+
+In Flutter, you can use stubs to replace real objects with mock objects or other simplified versions.
+
+To stub a method or class in Flutter, you can use the when() function provided by the `mockito` package. For example, if you have a method called myMethod() that you want to stub, you can do the following:
+
+```dart
+var myMock = MockMyObject();
+when(myMock.myMethod()).thenReturn('my result');
+```
+This will replace the `myMethod()` method on the `myMock` object with a stub that always returns the string 'my result'.
+
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
+
+## Talawa App Tests Example on How to use Mocks and Stubs Correctly
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests have all the necessary mocks and stubs to understand the concept of testing in talawa.
+
+### sendMessageToDirectChat Method
+
+```dart
+Future<void> sendMessageToDirectChat(
+    String chatId,
+    String messageContent,
+  ) async {
+    // trigger graphQL mutation to push the message in the Database.
+    final result = await _dbFunctions.gqlAuthMutation(
+      ChatQueries().sendMessageToDirectChat(),
+      variables: {"chatId": chatId, "messageContent": messageContent},
+    );
+
+    final message = ChatMessage.fromJson(
+      result.data['sendMessageToDirectChat'] as Map<String, dynamic>,
+    );
+
+    _chatMessageController.add(message);
+
+    debugPrint(result.data.toString());
+  }
+```
+
+Test written for this method looks like this
```

Comment:
There should be a title here like "Sample Mock and Test Code" with the text below.